### PR TITLE
feat(connext): multi currency support

### PR DESCRIPTION
**Pull request by ghost**
This PR adds support for withdrawing from on-chain address. In the process, `ethers` dependency is added to make development flow with the eth provider considerably faster. Previously, we were using the json-rpc API directly.

I'm planning to extract Connext swap client into a separate module so that the dependency will be opt-in (only for those who activate the swap client).

### CODE REVIEW

File: lib/connextclient/ConnextClient.ts
Comment (michael1011) [inline lib/connextclient/ConnextClient.ts:1106]: Should we really use the broadcast terminology? Why not just `on-chain transfer sent, transaction hash: ` and `could not send onchain transfer`?

File: lib/connextclient/ethprovider.ts
Comment (michael1011) [inline lib/connextclient/ethprovider.ts:43]: Why do we need the Ether balance for sending ERC20 tokens?

File: lib/connextclient/ethprovider.ts
Comment (michael1011) [inline lib/connextclient/ethprovider.ts:40]: Call my old fashioned but I like classes with constructors that initialize objects like the `provider` and simple function that only have a couple arguments. But you do you :shrug:

File: lib/connextclient/ConnextClient.ts
Comment (michael1011) [inline lib/connextclient/ConnextClient.ts:1099]: Aren't you missing the `units: string` argument here?

File: lib/connextclient/ethprovider.ts
Comment (michael1011) [inline lib/connextclient/ethprovider.ts:5]: Ethers has great TypeScript support in my experience. They don't expose a type for this?

**Review by michael1011 — COMMENTED** [review]
Concept ACK. Just some minor comments form my side

File: lib/connextclient/ethprovider.ts
Comment (erkarl) [reply lib/connextclient/ethprovider.ts:43]: This in preparation for adding balance verification logic that currently resides in `ConnextClient.withdraw`.

File: lib/connextclient/ethprovider.ts
Comment (erkarl) [reply lib/connextclient/ethprovider.ts:40]: I'm planning to refactor this function so that it receives signer (with provider) as an argument.

File: lib/connextclient/ConnextClient.ts
Comment (erkarl) [reply lib/connextclient/ConnextClient.ts:1099]: Yes, indeed :)

File: lib/connextclient/ethprovider.ts
Comment (erkarl) [reply lib/connextclient/ethprovider.ts:5]: No, since it's a contract function.
```
readonly [key: string]: ContractFunction | any;
```

File: lib/connextclient/ConnextClient.ts
Comment (erkarl) [reply lib/connextclient/ConnextClient.ts:1106]: I was thinking we might want to wait for the transfer event for emitting `complete` and `next` when broadcasted. But since complete is not implemented, yet - I'll change this to sent terminology.

Comment (erkarl) [reply lib/connextclient/ConnextClient.ts]: Changed.

File: lib/connextclient/ethprovider.ts
Comment (erkarl) [reply lib/connextclient/ethprovider.ts]: I went ahead and removed it right now to keep things easier. The checks are being done before this function is called.

File: lib/connextclient/ConnextClient.ts
Comment (michael1011) [reply lib/connextclient/ConnextClient.ts]: Just fyi: those events are emitted when the transaction is included in a block and not when it is accepted in the mempool

Comment (erkarl) [reply lib/connextclient/ConnextClient.ts]: Yup - good to know. We don't really care about that right now. The swapclient interface will expect `Promise<transactionId: string>`.

File: lib/connextclient/ethprovider.ts
Comment (michael1011) [reply lib/connextclient/ethprovider.ts:5]: What about [ContractTransaction](https://github.com/ethers-io/ethers.js/blob/master/packages/contracts/src.ts/index.ts#L90) or one of the types it extends from?

Comment (erkarl) [reply lib/connextclient/ethprovider.ts]: Yeap, good idea. Changed to `ContractTransactions` for the transfer call, but for balance calls I'm sticking with `Observable<ethers.BigNumber>` type since it's more accurate.

**Comment by erkarl** [comment]
This PR has evolved into:
- adding support for `xucli walletwithdraw` for ETH/ERC20
- monitoring channel for ERC20 deposits -> creating a new commitment transaction so that ERC20 tokens show up under channel balance"

Everything in the `ethprovider.ts` will be converted to a separate module. This in preparation for extracting `ConnextClient` into a separate module.

The PR is reviewable commit-by-commit so I'm not squashing the commits, yet.

File: jest.config.js
Comment (michael1011) [inline jest.config.js:7]: Why blank line?

File: test/jest/Connext.spec.ts
Comment (michael1011) [inline test/jest/Connext.spec.ts:120]: Why are we not allowing custom fees?

**Review by michael1011 — COMMENTED** [review]
utACK

Fixing those lint warnings would be nice though

File: test/jest/Connext.spec.ts
Comment (erkarl) [reply test/jest/Connext.spec.ts:120]: Not implemented, yet. This is porting over the functionality that we had for V1 to V2. We'll of course support this in the future.

File: jest.config.js
Comment (erkarl) [reply jest.config.js]: Removed. Thanks.

**Comment by sangaman** [comment]
> utACK
> 
> Fixing those lint warnings would be nice though

I left those warnings in when working on the eslint rules, they weren't trivial to fix but they also seemed like things that made sense to change (like redeclaring variables, which can be confusing and lead to some hard to find bugs imo), so I figured the warnings would be a good reminder to address later and to prevent new code from violating the rule. I didn't realize how annoying they'd be in the PR code diffs, but I guess that's a good motivator to actually fix these soonish.

That said, it doesn't look like this PR is introducing any new warnings, so I wouldn't let that hold this up.

File: .eslintrc.js
Comment (sangaman) [inline .eslintrc.js:9]: What is this for? It doesn't seem to cause any linting issues if I remove the line.

File: jest.config.js
Comment (sangaman) [inline jest.config.js:3]: I think this would conflict with the npm scripts which specify `testMatch` on a per-npm-script basis but share this same jest config file. Were you having a problem running jest tests?

**Review by sangaman — COMMENTED** [review]
Looks good so far. I'm not familiar with ramda (but I take it it will move out of xud eventually) and only superficially familiar with currying functions so it's not exactly apparent to me what's going on in a few places. I think more comments/method block comments in the ethProvider file would be good to explain what everything is does, so it's more comprehensible in the future for others who might not be so familiar either.

File: jest.config.js
Comment (erkarl) [reply jest.config.js:6]: `tsConfig` is deprecated in favour of `tsconfig`. After this change the deprecation warning disappeared.

File: .eslintrc.js
Comment (erkarl) [reply .eslintrc.js]: Indeed, think it was something related to my environment. Can't get any linting diff when removing it. Removed the change.

File: jest.config.js
Comment (erkarl) [reply jest.config.js]: > Were you having a problem running jest tests?

Indeed. It was the `npm run test:jest:watch` script that needed `testMatch` params. Removed this line `jest.config.js`

File: jest.config.js
Comment (sangaman) [reply jest.config.js:5]: Ah yes after some more googling it looks like you're right: https://github.com/kulshekhar/ts-jest/pull/2127

If you want to leave this as a separate commit when you're squashing (perhaps along with the change to `npm run test:jest:watch`) I think that would make sense.

**Comment by sangaman** [comment]
Let me know if/when I should review this again @erkarl .

**Comment by erkarl** [comment]
> Let me know if/when I should review this again @erkarl .

Will do. I'm waiting for a new Connext release to test that everything is working correctly in this branch.

**Comment by erkarl** [comment]
> That said, it doesn't look like this PR is introducing any new warnings, so I wouldn't let that hold this up.

Think it's not worth addressing them right now because once we've extracted `ConnextClient` into a separate module all code from current `ConnextClient.ts` should be ported over. Provided I don't introduce new warnings - these warning should disappear.

File: test/jest/Connext.spec.ts
Comment (erkarl) [reply test/jest/Connext.spec.ts:120]: Created an issue for this https://github.com/ExchangeUnion/xud/issues/2046

**Comment by erkarl** [comment]
> I think more comments/method block comments in the ethProvider file would be good to explain what everything is does, so it's more comprehensible in the future for others who might not be so familiar either.

@sangaman I added comments to ethProvider file - hope it is clearer now. Think this is ready for review again. I won't be adding more functionality to this PR.

**Review by michael1011 — APPROVED** [review]
utACK

File: lib/connextclient/ethprovider.ts
Comment (sangaman) [inline lib/connextclient/ethprovider.ts:88]: The smallest of nits, I'm seeing mixed casing of `Ethprovider` and `EthProvider`. I don't know which one is "correct" but might make sense to stick to one.

**Review by sangaman — APPROVED** [review]
Looks really good and comments are very helpful, thanks.

File: lib/connextclient/ethprovider.ts
Comment (erkarl) [reply lib/connextclient/ethprovider.ts:88]: Thanks. Will address this in a follow-up PR.

**Pull request merged**
